File: src/main/java/com/atsuishio/superbwarfare/capability/energy/SyncedEntityEnergyStorage.java
package com.atsuishio.superbwarfare.capability.energy;

import net.fabricmc.fabric.api.transfer.v1.transaction.TransactionContext;
import net.minecraft.nbt.Tag;
import net.minecraft.nbt.NumericTag;
import net.minecraft.network.syncher.EntityDataAccessor;
import net.minecraft.network.syncher.SynchedEntityData;

public class SyncedEntityEnergyStorage extends DynamicEnergyStorage {

    protected SynchedEntityData entityData;
    protected EntityDataAccessor<Integer> energyDataAccessor;

    public SyncedEntityEnergyStorage(long capacity, SynchedEntityData data, EntityDataAccessor<Integer> energyDataAccessor) {
        this(capacity, capacity, capacity, data, energyDataAccessor);
    }

    public SyncedEntityEnergyStorage(long capacity, long maxInsert, long maxExtract, SynchedEntityData data, EntityDataAccessor<Integer> energyDataAccessor) {
        super(() -> capacity, () -> maxInsert, () -> maxExtract);
        this.entityData = data;
        this.energyDataAccessor = energyDataAccessor;
    }

    @Override
    public long insert(long maxAmount, TransactionContext transaction) {
        this.amount = entityData.get(energyDataAccessor);
        long inserted = super.insert(maxAmount, transaction);
        if (inserted > 0) {
            entityData.set(energyDataAccessor, (int) this.amount);
        }
        return inserted;
    }

    @Override
    public long extract(long maxAmount, TransactionContext transaction) {
        this.amount = entityData.get(energyDataAccessor);
        long extracted = super.extract(maxAmount, transaction);
        if (extracted > 0) {
            entityData.set(energyDataAccessor, (int) amount);
        }
        return extracted;
    }

    @Override
    public long getAmount() {
        return entityData.get(energyDataAccessor);
    }

    @Override
    protected void readSnapshot(Long snapshot) {
        super.readSnapshot(snapshot);
        entityData.set(energyDataAccessor, snapshot.intValue());
    }

    public void deserializeNBT(Tag nbt) {
        if (nbt instanceof NumericTag numericTag) {
            setEnergy(numericTag.getAsInt());
        }
    }

    public void setEnergy(int energy) {
        this.amount = energy;
        entityData.set(energyDataAccessor, energy);
    }

    public void setCapacity(long capacity) {
        this.capacityGetter = () -> capacity;
    }

    public void setMaxExtract(long maxExtract) {
        this.maxExtractGetter = () -> maxExtract;
    }

    public void setMaxReceive(long maxReceive) {
        this.maxInsertGetter = () -> maxReceive;
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/capability/energy/SyncedEntityEnergyStorage.kt
package com.atsuishio.superbwarfare.capability.energy

import com.atsuishio.superbwarfare.capability.api.EnergyStorage
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.IntTag
import net.minecraft.nbt.Tag
import net.minecraft.network.syncher.EntityDataAccessor
import net.minecraft.network.syncher.SynchedEntityData

/**
 * 自动同步的实体能量存储能力，会和客户端自动同步实体的当前能量值
 */
open class SyncedEntityEnergyStorage(
    capacity: Int,
    maxReceive: Int,
    maxExtract: Int,
    protected var entityData: SynchedEntityData,
    protected var energyDataAccessor: EntityDataAccessor<Int>
) : EnergyStorage(capacity.toLong(), maxReceive.toLong(), maxExtract.toLong(), 0L) {

    constructor(
        capacity: Int,
        data: SynchedEntityData,
        energyDataAccessor: EntityDataAccessor<Int>
    ) : this(
        capacity,
        capacity,
        capacity,
        data,
        energyDataAccessor
    )

    init {
        // 从entityData同步初始能量值，避免reviveCaps()后内部energy字段与entityData不一致
        this.energy = entityData.get(energyDataAccessor).toLong()
    }

    fun setEnergy(energy: Int) {
        this.energy = energy.toLong()
        entityData.set(energyDataAccessor, energy)
    }

    fun setCapacity(capacity: Int) {
        this.capacity = capacity.toLong()
    }

    fun setMaxExtract(maxExtract: Int) {
        this.maxExtract = maxExtract.toLong()
    }

    fun setMaxReceive(maxReceive: Int) {
        this.maxReceive = maxReceive.toLong()
    }

    override fun onFinalCommit() {
        entityData.set(energyDataAccessor, this.energy.toInt())
    }

    override fun getAmount(): Long {
        // 获取同步数据，保证客户端能正确获得能量值
        return entityData.get(energyDataAccessor).toLong()
    }

    override fun deserializeNBT(provider: HolderLookup.Provider, nbt: Tag) {
        if (nbt is IntTag) {
            this.energy = nbt.asInt.toLong()
        } else {
            super.deserializeNBT(provider, nbt)
        }

        entityData.set(energyDataAccessor, this.energy.toInt())
    }
}
